Media/RTShaderLib/SGXLib_NormalMap.glsl

```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
 }
 
 //-----------------------------------------------------------------------------
-void SGX_TransformNormal(in vec3 vNormal,
-				         in vec4 vTangent,
+void SGX_CalculateTBN(in vec3 vNormal,
+					  in vec4 vTangent,
+					  out mat3 TBN)
+{	
+	vec3 vBinormal = cross(vNormal, vTangent.xyz) * vTangent.w;
+	
+	// direction: from tangent space to world
+	TBN = mtxFromCols(vTangent.xyz, vBinormal, vNormal);
+}
+
+//-----------------------------------------------------------------------------
+void SGX_CalculateTerrainTBN(in vec3 normal, in mat3 normalMatrix, out mat3 TBN)
+{
+    // derive the tangent space basis
+    // we do this in the pixel shader because we don't have per-vertex normals
+    // because of the LOD, we use a normal map
+    // tangent is always +x or -z in object space depending on alignment
+	#ifdef TERRAIN_ALIGN_Y_Z
+		vec3 tangent = vec3(0, 0, -1);
+	#else
+		vec3 tangent = vec3(1, 0, 0);
+	#endif
+    normal = normalize(normal);
+    tangent = normalize(mul(normalMatrix, tangent));
+    vec3 binormal = cross(tangent, normal);
+    // note, now we need to re-cross to derive tangent again because it wasn't orthonormal
+    tangent = cross(normal, binormal);
+    // derive final matrix
+    TBN = mtxFromCols(tangent, binormal, normal);
+}
+					  
+//-----------------------------------------------------------------------------
+void SGX_TransformNormal(in mat3 TBN,
 						 inout vec3 vNormalTS)
 {
-	// use non-normalised post-interpolation values as in mikktspace
-	// resulting normal will be normalised by lighting
-	vec3 vBinormal = cross(vNormal, vTangent.xyz) * sign(vTangent.w);
-	// direction: from tangent space to world
-	mat3 TBN = mtxFromCols(vTangent.xyz, vBinormal, vNormal);
 	vNormalTS = mul(TBN, vNormalTS);
 }
 
 //-----------------------------------------------------------------------------
 void SGX_Generate_Parallax_Texcoord(in sampler2D normalHeightMap,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in mat3 TBN,
+						out vec2 newTexCoord)
+{
+	//Calculate eye direction
+	TBN = transpose(TBN);
+	vec3 eyeVec = mul(TBN, -viewPos);
```

Comment:
```suggestion
	vec3 eyeVec = mul(-viewPos, TBN); // transpose(TBN) * -viewPos
```

Reply:
I tried this but unfortunately it breaks the parallax effect - it looks very flat with this change.

Reply:
Oh no sorry - it does work - didn't remove TBN = transpose.